Add tslint-plugin-prettier and apply code formatting

mturley opened the pull request:
<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What issue is being addressed here?) -->

**What**:

Because we realized that [some important code-formatting-related TSLint rules are broken](https://github.com/patternfly/patternfly-react/pull/2787#issuecomment-526628618), @redallen and I decided that it would be appropriate to add [Prettier](https://prettier.io/) to our linter, and run it on existing code.

This PR:
* Adds `tslint-plugin-prettier` which runs Prettier as part of TSLint (currently these errors do not fail the CI checks, but eventually when other TSLint errors are resolved they probably should)
* Re-adds `tslint-config-prettier` which was removed in #2787, this disables TSLint rules which would conflict with Prettier
* Re-adds the `yarn prettier` script which contributors can run to automatically format their code
* Includes code changes from an initial run of `yarn prettier` and `yarn lint:ts --fix`. This cleans up all of our indentation issues and other small formatting gripes.

Hopefully this will help to make our code a little more readable and consistent going forward!

cc @redallen 

Closes https://github.com/patternfly/patternfly-react/issues/2362 (for real this time)

mturley commented:
Hmm, `yarn lint:ts --fix` still makes some conflicting changes. Hang on for one more commit.

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2810.surge.sh

mturley commented:
Ok, this should be all set now. I ran `yarn lint:ts --fix` and it had some more changes to make, then ran `yarn prettier` again and it made no changes, so they don't conflict.

mturley commented:
Rebased on master

redallen commented:
Let's try to do get this in post code-freeze next week after one final rebase. Things are a little hectic now.

mturley commented:
Fair enough 🙂

mturley commented:
I rebased this on the latest master, and squashed it into two commits to make it easier to rebase in case we need to do so again:

* https://github.com/patternfly/patternfly-react/pull/2810/commits/a56c6b88578b047b9bab16e2f4ed37a00aa328a3 makes the actual configuration changes, this commit should be included in a rebase
* https://github.com/patternfly/patternfly-react/pull/2810/commits/51a85c05483fa40044710c934de9124742ed6d1e runs the linter and formatter on existing code, this commit should be dropped and replaced in a rebase

mturley commented:
Hmm.. @redallen I'm having trouble parsing the errors in the CircleCI log, do you know why this would be failing CI? I didn't make any different configuration changes compared to before the rebase.

redallen commented:
@mturley During `yarn build` it's something to do with Chart types:
```
src/components/ChartVoronoiContainer/ChartVoronoiContainer.tsx(153,6): error TS2322: Type '{ activateData?: boolean; activateLabels?: boolean; disable?: boolean; labels?: (point: any, index: number, points: any[]) => string; mouseFollowTooltips?: boolean; onActivated?: (points: any[], props: VictoryVoronoiContainerProps) => void; ... 14 more ...; theme: ChartThemeDefinitionInterface; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<VictoryVoronoiContainer> & Readonly<VictoryVoronoiContainerProps> & Readonly<{ children?: ReactNode; }>'.
  Property 'className' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<VictoryVoronoiContainer> & Readonly<VictoryVoronoiContainerProps> & Readonly<{ children?: ReactNode; }>'
```

Charts has been touched a lot lately, so I'd recheck the merge conflicts for that file. Worst case, checkout master's version and then relint the file.

mturley commented:
> Worst case, checkout master's version and then relint the file
Hmm, that's what I did. I'll take a closer look at that file, thanks!

mturley commented on packages/patternfly-4/react-charts/src/components/ChartVoronoiContainer/ChartVoronoiContainer.tsx:
```diff
@@ -143,16 +140,18 @@ export const ChartVoronoiContainer: React.FunctionComponent<ChartVoronoiContaine
   // Note: theme is required by voronoiContainerMixin, but @types/victory is missing a prop type
   // @ts-ignore
-  return <VictoryVoronoiContainer className={chartClassName} labelComponent={chartLabelComponent} theme={theme} {...rest} />;
+  return (
```
Ahh, @redallen it looks like it was this `@ts-ignore` not being applied anymore because the offending code was moved down a line by prettier. I'll add a `prettier-ignore` on the same line, and we might have to do that in the future where we need `@ts-ignore` on a long line.

It's weird that we can't just `ts-ignore` multiple lines... looks like an open issue: https://github.com/Microsoft/TypeScript/issues/19573

mturley commented:
I rebased this on the latest master, and squashed it into fewer commits to make it easier to rebase in case we need to do so again:

* https://github.com/patternfly/patternfly-react/pull/2810/commits/61fe744dd9cb868e3526ab7f078e53b427ba4a4b makes the actual configuration changes, this commit should be included in a rebase
* https://github.com/patternfly/patternfly-react/pull/2810/commits/de1e7e5ec6ee32be25c6328cdb0847d4f1c9f806 stops prettier from making the build fail by removing ts-ignore from a line, this commit should be included in a rebase
* https://github.com/patternfly/patternfly-react/pull/2810/commits/e771fff6067e40866d37ce9b0400b63cb74ee8af runs the linter and formatter on existing code, this commit should be dropped and replaced in a rebase

(edited 2x)

mturley commented:
@dgutride do you think it would make sense to try and merge this (after a final rebase) immediately after this week's code freeze ends, to minimize conflicts with other PRs? It doesn't change our dist, but it touches a lot of files.

mturley commented:
Rebased and resolved conflicts with https://github.com/patternfly/patternfly-react/pull/2895

mturley commented:
@dlabrecq what formatting opinions of Prettier do you dislike? We can reconfigure it: https://prettier.io/docs/en/options.html. I'm not really concerned about how the code is formatted as long as it's formatted consistently. Messy indentation was my main gripe.

mturley commented:
I do agree somewhat that Prettier tries too much to fit the most code into every line, but I think maybe that could be helped by lowering the print-width option. They actually recommend that you don't exceed 80 characters on that option (see the yellow box at the top of the [options docs](https://prettier.io/docs/en/options.html), and we have it set to 120.

I would be open to shortening this to 80 or 100 (and then obviously there would be many lines which would end up longer than that).

dlabrecq commented:
@mturley I often use separate lines when I destructure multiple items or sometimes when importing long names, but that's just my preference to make things more readable. It's not a big deal, just not a fan. I am a big fan of running `yarn lint:ts --fix`, tho

mturley commented:
I agree actually, those sound like good reasons to disagree with Prettier. We could lower the print width to make that less of an issue, and in extreme cases we could use `// prettier-ignore` if it's being really fussy. This PR doesn't automatically run it going forward though, so we can always run it and then change things back if we disagree with it.

mturley commented:
@dlabrecq I'll bug you just one last time (I know you're not thinking of this as a big deal), but just for comparison's sake I made a few separate branches and ran the same process with the print-width set to 100 characters (https://github.com/mturley/patternfly-react/commit/c5fbfe4a62f6fcb0f1a5a7517da20d5b7c782325) and set to 80 characters (https://github.com/mturley/patternfly-react/commit/fc02c3ba249931e780a874bfd3fdaa4cfe14282d).

The lower the print-width, the more readable a lot of code gets, but files end up larger in terms of lines of code. Do you have an opinion on whether either of these would be better than the 120-character width in this PR?

dlabrecq commented on packages/patternfly-4/react-charts/src/components/ChartBullet/utils/chart-bullet-size.ts:
```diff
@@ -121,41 +126,34 @@ export const getQualitativeRangeBarWidth = ({
+    ? height > defaultSize
+      ? value + (height - defaultSize) * scale
+      : value - (defaultSize - height) * scale
+    : width > defaultSize
```
This indentation looks much better compared to the others I noted

dlabrecq commented on packages/patternfly-4/react-charts/src/components/ChartPoint/ChartPoint.tsx:
```diff
@@ -75,8 +75,18 @@ export interface ChartPointProps {
+    | 'triangleUp'
+    | 'dash'
+    | 'threshold'
+    | Function;
```
Better...

dlabrecq commented on packages/patternfly-4/react-charts/src/components/ChartBullet/ChartBullet.tsx:
```diff
@@ -720,12 +715,16 @@ export const ChartBullet: React.FunctionComponent<ChartBulletProps> = ({
+        ? defaultPadding.top * 0.5 + (defaultPadding.bottom * 0.5 - defaultPadding.bottom) - 25
+        : title
+        ? -defaultPadding.bottom + 60
+        : -defaultPadding.bottom;
```
This indentation looks a little off?

mturley replied:
It looks like Prettier will indent a nested ternary if it is nested within the true-branch (`foo ? bar ? 1 : 2 : 3`) but not if it is trailing / nested in the false-branch (`foo ? 1 : bar ? 2 : 3`). This is a bit weird but seems reasonable to me... to be honest though, maybe we should avoid nested ternary expressions unless we use parentheses to make them more clear. Maybe that would help here.

mturley replied:
Looks like this is a subject of debate in Prettier: https://github.com/prettier/prettier/issues/5814

mturley replied:
Looks like this is a subject of debate in Prettier: prettier/prettier#5814

mturley replied:
Huh. Adding parens around the second expression there, Prettier removes them. Maybe that's a case where `// prettier-ignore` would be useful if we need an exception to make that code extra clear.

dlabrecq replied:
I didn't realize `// prettier-ignore` was available. That could be quite useful for those couple of weird issues

mturley replied:
Agreed. You can even configure a `.prettierignore` file to ignore entire files/directories if necessary

mturley commented:
Rebased, if you want to re-approve @jschuler @dlabrecq

mturley commented:
🎶 _...one more time..._ 🎶 

![giphy](https://user-images.githubusercontent.com/811963/65974410-eb64d200-e43a-11e9-8410-0ed799fc8822.gif)

patternfly-build commented:
Your changes have been released in: 
 - patternfly-react-extensions@2.20.9
 - patternfly-react-wooden-tree@2.0.5
 - patternfly-react@2.39.4
 - @patternfly/react-console@1.12.13
 - @patternfly/react-charts@5.0.15
 - @patternfly/react-core@3.112.4
 - @patternfly/react-docs@4.14.4
 - @patternfly/react-inline-edit-extension@2.11.71
 - demo-app-ts@3.6.11
 - @patternfly/react-integration@3.6.2
 - @patternfly/react-styled-system@3.6.40
 - @patternfly/react-styles@3.5.28
 - @patternfly/react-table@2.22.20
 - @patternfly/react-topology@2.8.66
 - @patternfly/react-virtualized-extension@1.2.56
 - @patternfly/react-icons@3.14.8

Thanks for your contribution! :tada:
